### ISSUE TITLE
False positive when using type set as function

### ISSUE THREAD

**Issue by darkclouder** [python/mypy#9455]
**Bug Report**

I get false positives (to be of wrong type) when I use the `set` constructor as a `Callable[[Iterable[T]], Set[T]]`, however only in some cases.
Whenever I start to add a function parameter, I get this false positive, even though it has nothing to do with the actual line where I get the error.

I played around with different variants and for some I got mypy errors, for others I didn't.
* `func_1` works.
* The moment I add another parameter like in `func_2` or `func_2b`, I get mypy errors.
* Passing data itself as a parameter also leads to the error, like in `func_3` or `func_3b`
* I tried to explicitly type `items` in `func_3b` and `func_4b`, but it seems like it can't even match the types when I hint it in the right direction.
* I wrapped `set` in a custom generic function because I hoped it might resolve it in `func_4` and `func_4b`, but it doesn't.
* Wrapping works just like the original set if I don't have a function parameter, see `func_5`.
* However: Wrapping `set` in a non-generic function, replacing `T` with `str` works: `func_6`.

Where I see a bit of a pitfall here is that `set` returns a `Set[X]`, which means if you substitute the type in `map_or_else` then `func` would be of type `Callable[[S], Set[X]]` and therefore `Set[X]=T`.
However, my parameter for default is always `None`, so `T` should also include `None`.
We conclude that `T=Optional[Set[X]]`. Which in turn means that `func` actually must be of type `Callable[[S], Optional[Set[X]]]`.
However, set is a function of type `Callable[[S], Set[X]]`.
I would argue that `Callable[[S], Optional[Set[X]]]` is covariant to `Callable[[S], Set[X]]`, so it should be fine IMO.
Note that changing `map_or_else`'s signature to `map_or_else(func: Callable[[S], T], value: Optional[S], default: Optional[T]) -> Optional[T]` did not resolve the issue.

**To Reproduce**

Run mpypy on this code:

```python
from typing import *


S = TypeVar("S")
T = TypeVar("T")


def map_or_else(func: Callable[[S], T], value: Optional[S], default: T) -> T:
    if value is None:
        return default
    return func(value)


# No mypy error
def func_1():
    data: Dict[str, List[str]] = {
        "items": [
            "a", "b", "c",
        ],
    }

    items = map_or_else(set, data.get("items"), None)

    if items is not None:
        print(items)



# Mypy error
def func_2(other_data: Dict[str, List[str]]):
    data: Dict[str, List[str]] = {
        "items": [
            "a", "b", "c",
        ],
    }

    items = map_or_else(set, data.get("items"), None)

    if items is not None:
        print(items)


# Mypy error
def func_2b(just_a_string: str):
    data: Dict[str, List[str]] = {
        "items": [
            "a", "b", "c",
        ],
    }

    items = map_or_else(set, data.get("items"), None)

    if items is not None:
        print(items)


# Mypy error
def func_3(data: Dict[str, List[str]]):
    items = map_or_else(set, data.get("items"), None)

    if items is not None:
        print(items)


# Mypy error
def func_3b(data: Dict[str, List[str]]):
    items: Optional[Set[str]] = map_or_else(set, data.get("items"), None)

    if items is not None:
        print(items)


def make_set(items: Iterable[T]) -> Set[T]:
    return set(items)


# Mypy error
def func_4(data: Dict[str, List[str]]):
    items = map_or_else(make_set, data.get("items"), None)

    if items is not None:
        print(items)


# Mypy error
def func_4b(data: Dict[str, List[str]]):
    make_set: Callable[[Iterable[T]], Set[T]] = set

    items: Optional[Set[str]] = map_or_else(make_set, data.get("items"), None)

    if items is not None:
        print(items)


# No mypy error
def func_5():
    data: Dict[str, List[str]] = {
        "items": [
            "a", "b", "c",
        ],
    }

    items = map_or_else(make_set, data.get("items"), None)

    if items is not None:
        print(items)


def make_set_2(items: Iterable[str]) -> Set[str]:
    return set(items)


# No mypy error
def func_6(data: Dict[str, List[str]]):
    items = map_or_else(make_set_2, data.get("items"), None)

    if items is not None:
        print(items)



if __name__ == "__main__":
    some_data: Dict[str, List[str]] = {
        "items": [
            "a", "b", "c",
        ],
    }

    func_1()
    func_2({})
    func_2b("")
    func_3(some_data)
    func_3b(some_data)
    func_4(some_data)
    func_4b(some_data)
    func_5()
    func_6(some_data)

```

**Expected Behavior**

None of the variants should give mypy errors.

**Actual Behavior**

Some of the variants give mypy errors.
Most notably with the rather arbitrary behaviour that the error only comes up when I add function parameters.

**Your Environment**

- Mypy version used: mypy 0.782
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.7.4, 3.8.5
- Operating system and version: macOS 10.15.4

**Comment by hhoppe** [comment]
I have a similar issue, which can be reproduced by a 1-line program:
```
map(set, ['ab', 'bc'])
```
`mypy` 0.991 reports:
```
error: Argument 1 to "map" has incompatible type "Type[Set[Any]]"; expected "Callable[[str], Set[_T]]"  [arg-type]
```

**Comment by ilevkivskyi** [comment]
This now works correctly.
